### PR TITLE
bluetooth: tester: GMCS Server tests

### PR DESCRIPTION
**Pull request by piotrnarajowski**
Add support for GMCS tests.

### CODE REVIEW

File: tests/bluetooth/tester/src/btp/btp.h [hunk 1]
```diff
@@ -61,8 +62,9 @@
 #define BTP_SERVICE_ID_VCP	20
 #define BTP_SERVICE_ID_CAS	21
 #define BTP_SERVICE_ID_MCP	22
+#define BTP_SERVICE_ID_GMCS	23
 
-#define BTP_SERVICE_ID_MAX	BTP_SERVICE_ID_MCP
+#define BTP_SERVICE_ID_MAX	BTP_SERVICE_ID_GMCS
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp/btp.h:67]: Are we going to have both `MCS` and `GMCS` here?

File: tests/bluetooth/tester/src/btp/btp_mcs.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+struct btp_mcs_media_state_ev {
+	uint8_t state;
+} __packed;
+
+#define BTP_MCS_COMMAND_EV			0x81
+struct btp_mcs_command_ev {
+	uint8_t status;
+	uint8_t opcode;
+	uint8_t  use_param;
+	int32_t param;
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp/btp_mcs.h:37]:
```suggestion
	uint8_t opcode;
	uint8_t use_param;
	int32_t param;
```

File: tests/bluetooth/tester/src/btp/btp_mcs.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+#define BTP_MCS_READ_SUPPORTED_COMMANDS		0x01
+struct btp_mcs_read_supported_commands_rp {
+	uint8_t data[0];
+} __packed;
+
+#define BTP_MCS_CMD_SEND			0x02
+struct btp_mcs_send_cmd {
+	uint8_t  opcode;
+	uint8_t use_param;
+	int32_t param;
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp/btp_mcs.h:20]:
```suggestion
	uint8_t opcode;
	uint8_t use_param;
	int32_t param;
```

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 4]
```diff
@@ -1317,3 +1320,224 @@ uint8_t tester_unregister_mcp(void)
...
+}
+
+static void btp_send_media_player_command_ev(uint8_t status, const struct mpl_cmd *cmd)
+{
+	struct btp_mcs_command_ev ev;
+
+	ev.status = status;
+	ev.opcode = cmd->opcode;
+	ev.use_param = cmd->use_param;
+	ev.param = sys_cpu_to_le32(cmd->param);
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_mcp.c:1398]: In the case that `use_param` is 0x00 (`false`), is `cmd->param` always initialized?

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bluetooth/tester/src/btp_mcp.c (hunk 4 above)
Comment (Thalley) [reply tests/bluetooth/tester/src/btp_mcp.c:1398]: Nah, just wanted to ensure we weren't calling `sys_cpu_to_le32` with uninitialized values :)

File: tests/bluetooth/tester/src/btp/btp.h (hunk 1 above)
Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp/btp.h]: Added BTP_SERVICE_ID_MCS (24) for future use.

File: tests/bluetooth/tester/src/btp/btp.h [hunk 5]
```diff
@@ -61,8 +62,10 @@
 #define BTP_SERVICE_ID_VCP	20
 #define BTP_SERVICE_ID_CAS	21
 #define BTP_SERVICE_ID_MCP	22
+#define BTP_SERVICE_ID_GMCS	23
+#define BTP_SERVICE_ID_MCS	24
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp/btp.h:66]: So in the test spec/test suite these 2 services with (near?) identical behavior will be split?

**Review by Thalley — DISMISSED** [review]

File: tests/bluetooth/tester/src/btp/btp.h (hunk 5 above)
Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp/btp.h:66]: Exactly, tests are split to GMCS and MCS with identical behavior.

**Review by sjanc — CHANGES_REQUESTED** [review]
this needs to be updated to BTP changes requested in https://github.com/auto-pts/auto-pts/pull/1054

File: tests/bluetooth/tester/src/btp/btp_mcs.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+#define BTP_MCS_READ_SUPPORTED_COMMANDS		0x01
+struct btp_mcs_read_supported_commands_rp {
+	uint8_t data[0];
+} __packed;
+
+#define BTP_MCS_CMD_SEND			0x02
+struct btp_mcs_send_cmd {
+	uint8_t opcode;
+	uint8_t use_param;
+	uint32_t param;
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp/btp_mcs.h:20]: The `param` in the MPL implementation is a `int32`. Should this be the same?

Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp/btp_mcs.h:20]: Ah yes, missed that.

**Review by Thalley — DISMISSED** [review]

**Comment by piotrnarajowski** [comment]
Updated regarding to @asbjornsabo comments on Media Player issues mentioned above

**Review by Thalley — DISMISSED** [review]

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 7]
```diff
@@ -1317,3 +1326,242 @@ uint8_t tester_unregister_mcp(void)
...
+	LOG_DBG("MCS Read Next Track Obj Id");
+
+	err = media_proxy_ctrl_get_next_track_id(mcs_media_player);
+	if (err) {
+		return BTP_STATUS_FAILED;
+	}
+
+	sys_put_le48(next_track_obj_id, rp->id);
+
+	*rsp_len = sizeof(*rp) + 1;
```
Comment (sjanc) [inline tests/bluetooth/tester/src/btp_mcp.c:1384]: this should be sizeof(*rp), shouldn't it?   (there is no variable part here)

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 8]
```diff
@@ -1317,3 +1326,242 @@ uint8_t tester_unregister_mcp(void)
...
+	LOG_DBG("MCS Read Current Track Obj Id");
+
+	err = media_proxy_ctrl_get_current_track_id(mcs_media_player);
+	if (err) {
+		return BTP_STATUS_FAILED;
+	}
+
+	sys_put_le48(current_track_obj_id, rp->id);
+
+	*rsp_len = sizeof(*rp) + 1;
```
Comment (sjanc) [inline tests/bluetooth/tester/src/btp_mcp.c:1404]: ditto

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 9]
```diff
@@ -1317,3 +1326,242 @@ uint8_t tester_unregister_mcp(void)
...
+{
+	struct btp_mcs_state_set_rp *rp = rsp;
+
+	LOG_DBG("MCS Set Media Player to inactive state");
+
+	mpl_test_media_state_set(MEDIA_PROXY_STATE_INACTIVE);
+
+	rp->state = media_player_state;
+
+	*rsp_len = sizeof(*rp) + 1;
```
Comment (sjanc) [inline tests/bluetooth/tester/src/btp_mcp.c:1447]: ditto

**Review by sjanc — CHANGES_REQUESTED** [review]

File: tests/bluetooth/tester/src/btp/btp.h (hunk 5 above)
Comment (sjanc) [inline tests/bluetooth/tester/src/btp/btp.h:66]: I'd not define BTP_SERVICE_ID_MCS until it is used, so just remove it

File: tests/bluetooth/tester/src/btp_mcp.c (hunk 7 above)
Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp_mcp.c]: Fixed

File: tests/bluetooth/tester/src/btp/btp.h (hunk 5 above)
Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp/btp.h]: Fixed

**Review by Thalley — APPROVED** [review]

**Review by sjanc — APPROVED** [review]

**Pull request merged**